core/devdoc/oop_module_requirements.md

```diff
@@ -0,0 +1,106 @@
+Module Requirements
+===================
+
+## Overview
+
+This is the documentation for an out-of-process module that communicates with
+other modules through the Azure IoT Field Gateway. Out-of-process modules need
+to implement the well defined Module Interface. Each out-of-process module will
+additionally need to attach and detach from the gateway and will either need to
+manage the worker thread manually or use the convenience methods supplied to
+start and halt the worker thread.
+
+## References
+
+[Azure IoT Field Gateway Module Interface](module.md)
+
+![High Level Design Diagrams](media/oop_module_hld.PNG)
+
+## Exposed Types
+
+```c
+typedef OOP_MODULE * OOP_MODULE_HANDLE;
+```
+
+## Exposed API
+
+### OopModule_Attach
+
+`OopModule_Attach` will provide the OopModule library with the caller's
+"Module API" and the connection key required to communicate with the Azure IoT
+Field Gateway. When `OopModule_Attach` is called the OopModule library will begin
+listening for the gateway create message. Once the create message is received
+the OopModule library will call `Module_ParseConfigurationFromJson`^, `Module_Create`
+and `Module_FreeConfiguration`^.
+*^ Assuming the optional API has been provided.*
+
+```c
+extern
+OOP_MODULE_HANDLE
+OopModule_Attach (
+    const MODULE_API * module_api_,
+    const char * const command_channel_uri_
+);
+```
+
+### OopModule_Detach
+
+`OopModule_Detach` will cease to interact with the Azure IoT Field Gateway
+and clean up all resources required to maintain the connection with the gateway.
+
+```c
+extern
+void
+OopModule_Detach (
+    OOP_MODULE_HANDLE oop_module_
+);
+```
+
+### OopModule_DoWork
+
+`OopModule_DoWork` is intended to provide the caller with fine-grain control of the
+worker thread. Therefore, `OopModule_DoWork` is a non-blocking call, which will check
```

Comment:
@damonbarry As it stands, `DoWork` only receives messages. Broker_Publish is intercepted and directly invokes nanomsg, so it falls outside of `DoWork`. Do you have a vision for how this would work? Feels like a queueing mechanism.